rfcs/0126-standardize-magic-properties.md

```diff
@@ -0,0 +1,81 @@
+---
+feature: standardize-magic-properties
+start-date: 2022-05-26
+author: Anselm Schüler
+co-authors: None
+shepherd-team: None
+shepherd-leader: None
+related-issues: None
+---
+
+# Summary
+[summary]: #summary
+
+Rename attrset attribute names treated specially by Nix to start with a double underscore.
+Add a builtin function `guardMagic` that returns `null` for double underscore prefixed strings.
+
+# Motivation
+[motivation]: #motivation
+
+Nix treats many attributes specially with unique behaviours. Some of these are prefixed with double underscores to disambiguate them from everything else. But some aren’t: `recurseForDerivations` isn’t, `type`, when set to `"derivation"`, changes visual output.
```

Comment:
I have a hard time understanding the exact problem and its severity.

Could you please make an explicit list of the magic attributes you are aware of, explain what they are for, why they are magic, and link to relevant documentation and source code?

Reply:
The problem is that there’s a bunch of specially treated attribute names that mean attrsets aren’t neutral data structures.
Maybe magic is a bad word.
As I say further down, I don’t know a full list, since these seem to be things that aren’t considered together in the documentation.

`type` is magic because if `type = "derivation"`, the data is displayed completely differently in the Nix repl.
By renaming them to start with `__`, I hope to make them more obvious, more segregated from normal attributes, and easier to deal with.

A problem this could fix could arise when a user who does not know how a process works inputs data into it, which is unknowingly processed to create an attribute that unexpectedly alters the result’s behaviour.
None exist ATM AFAIK, but they could.

Reply:
Attrsets by themselves are (almost?) neutral, I believe, but some functions consuming them treat some attributes specially.

Reply:
With the exception of `__functor`. And precisely these, in my opinion, should be dunder(__)-prefixed

Reply:
> type is magic because if type = "derivation", the data is displayed completely differently in the Nix repl.
By renaming them to start with __, I hope to make them more obvious, more segregated from normal attributes, and easier to deal with.

[it's even more magic than that](https://github.com/NixOS/nix/blob/a72e75e6b6db5506f0932d8f1d507cdf3a466130/src/libexpr/eval.cc#L2207). changing `type` is a flag day event, and for that reason alone probably not worth it. also note that `recurseForDerivations` isn't (afawk( honored by the new nix commands at all either.

Reply:
That’s a good point. Looking at the code though it seems not so hard to change. Don’t the new commands try to build non-derivation store path results anyways? But yeah, this would be a big change, probably not worth it. Also, AFAIK nix-command honors recurseForDerivations in legacyPackages *only*.